File: windows/src/storage_manager.py
"""
storage_manager.py — File I/O, directory management, and media index persistence.
"""
import json
import os
import threading
from dataclasses import asdict, dataclass
from datetime import datetime
from pathlib import Path
from typing import Optional

from app_settings import AppSettings


# ---------------------------------------------------------------------------
# Media item model (stored in media_index.json)
# ---------------------------------------------------------------------------

@dataclass
class MediaItem:
    profile_username: str
    media_type: str             # Posts / Reels / Videos / Highlights / Stories / Profile Pictures
    instagram_id: str
    media_url: str
    local_path: Optional[str]
    caption: Optional[str]
    timestamp: str              # ISO8601 string
    downloaded_at: str          # ISO8601 string
    file_size: Optional[int]    # bytes

    def to_dict(self) -> dict:
        return asdict(self)

    @staticmethod
    def from_dict(d: dict) -> "MediaItem":
        return MediaItem(
            profile_username=d.get("profile_username", ""),
            media_type=d.get("media_type", "Posts"),
            instagram_id=d.get("instagram_id", ""),
            media_url=d.get("media_url", ""),
            local_path=d.get("local_path"),
            caption=d.get("caption"),
            timestamp=d.get("timestamp", datetime.now().isoformat()),
            downloaded_at=d.get("downloaded_at", datetime.now().isoformat()),
            file_size=d.get("file_size"),
        )


# ---------------------------------------------------------------------------
# StorageManager
# ---------------------------------------------------------------------------

MEDIA_TYPES = [
    "Posts", "Reels", "Videos", "Highlights", "Stories", "Profile Pictures"
]


class StorageManager:
    _instance = None
    _lock = threading.Lock()

    def __new__(cls):
        with cls._lock:
            if cls._instance is None:
                cls._instance = super().__new__(cls)
                cls._instance._initialized = False
        return cls._instance

    def __init__(self):
        if self._initialized:
            return
        self._settings = AppSettings()
        self._index_lock = threading.Lock()
        self._initialized = True

    # ------------------------------------------------------------------
    # Paths
    # ------------------------------------------------------------------

    @property
    def _index_path(self) -> Path:
        return self._settings.app_data_dir / "media_index.json"

    @property
    def _profiles_path(self) -> Path:
        return self._settings.app_data_dir / "profiles.json"

    def profile_dir(self, username: str) -> Path:
        return Path(self._settings.download_path) / username

    def media_dir(self, username: str, media_type: str) -> Path:
        return self.profile_dir(username) / media_type

    def preferred_extension(self, media_url: Optional[str], fallback_is_video: bool) -> str:
        if media_url:
            lowered = media_url.lower()
            path = Path(media_url.split("?", 1)[0].split("#", 1)[0])
            ext = path.suffix.lower().lstrip(".")
            if ext in {"jpg", "jpeg", "png", "webp", "mp4", "mov", "m4v"}:
                if ext == "jpeg":
                    return "jpg"
                if ext in {"mov", "m4v"}:
                    return "mp4"
                return ext
            if ".mp4" in lowered or "video" in lowered:
                return "mp4"
            if ".webp" in lowered:
                return "webp"
            if ".png" in lowered:
                return "png"
        return "mp4" if fallback_is_video else "jpg"

    def detected_extension(self, data: bytes, media_url: Optional[str], fallback_is_video: bool) -> str:
        if len(data) >= 12 and data[4:8] == b"ftyp":
            return "mp4"
        if len(data) >= 3 and data[:3] == b"\xFF\xD8\xFF":
            return "jpg"
        if len(data) >= 8 and data[:8] == b"\x89PNG\r\n\x1a\n":
            return "png"
        if len(data) >= 12 and data[:4] == b"RIFF" and data[8:12] == b"WEBP":
            return "webp"
        return self.preferred_extension(media_url, fallback_is_video)

    def save_path(self, username: str, media_type: str, instagram_id: str,
                  timestamp: datetime, is_video: bool, index: int = 0,
                  total: int = 1, media_url: Optional[str] = None,
                  file_ext: Optional[str] = None) -> Path:
        ext = (file_ext or self.preferred_extension(media_url, is_video)).lower()
        ts  = int(timestamp.timestamp())
        suffix = f"_{index + 1}" if total > 1 else ""
        filename = f"{instagram_id}_{ts}{suffix}.{ext}"
        return self.media_dir(username, media_type) / filename

    def existing_save_path(self, username: str, media_type: str, instagram_id: str,
                           timestamp: datetime, is_video: bool, index: int = 0,
                           total: int = 1, media_url: Optional[str] = None) -> Optional[Path]:
        preferred = self.save_path(
            username,
            media_type,
            instagram_id,
            timestamp,
            is_video,
            index,
            total,
            media_url=media_url,
        )
        if preferred.exists():
            return preferred

        ts = int(timestamp.timestamp())
        suffix = f"_{index + 1}" if total > 1 else ""
        base = f"{instagram_id}_{ts}{suffix}"
        for ext in ("mp4", "jpg", "jpeg", "png", "webp"):
            candidate = self.media_dir(username, media_type) / f"{base}.{ext}"
            if candidate.exists():
                return candidate
        return None

    # ------------------------------------------------------------------
    # Directory creation
    # ------------------------------------------------------------------

    def ensure_dir(self, path: Path) -> Path:
        path.mkdir(parents=True, exist_ok=True)
        return path

    def create_profile_directories(self, username: str):
        for mt in MEDIA_TYPES:
            self.ensure_dir(self.media_dir(username, mt))

    # ------------------------------------------------------------------
    # Profile persistence
    # ------------------------------------------------------------------

    def save_profiles(self, profiles: list[dict]):
        try:
            with open(self._profiles_path, "w", encoding="utf-8") as f:
                json.dump(profiles, f, indent=2, default=str)
        except Exception as e:
            print(f"[Storage] Failed to save profiles: {e}")

    def load_profiles(self) -> list[dict]:
        if not self._profiles_path.exists():
            return []
        try:
            with open(self._profiles_path, "r", encoding="utf-8") as f:
                return json.load(f)
        except Exception as e:
            print(f"[Storage] Failed to load profiles: {e}")
            return []

    # ------------------------------------------------------------------
    # Media index
    # ------------------------------------------------------------------

    def save_media_index(self, items: list[MediaItem]):
        with self._index_lock:
            try:
                data = [i.to_dict() for i in items]
                tmp = self._index_path.with_suffix(".tmp")
                with open(tmp, "w", encoding="utf-8") as f:
                    json.dump(data, f)
                tmp.replace(self._index_path)
            except Exception as e:
                print(f"[Storage] Failed to save media index: {e}")

    def load_media_index(self) -> list[MediaItem]:
        if not self._index_path.exists():
            return []
        try:
            with open(self._index_path, "r", encoding="utf-8") as f:
                data = json.load(f)
            return [MediaItem.from_dict(d) for d in data]
        except Exception as e:
            print(f"[Storage] Failed to load media index: {e}")
            return []

    # ------------------------------------------------------------------
    # File operations
    # ------------------------------------------------------------------

    def save_file(self, data: bytes, path: Path):
        self.ensure_dir(path.parent)
        tmp = path.with_suffix(".tmp")
        with open(tmp, "wb") as f:
            f.write(data)
        tmp.replace(path)

    def delete_profile_files(self, username: str):
        import shutil
        profile_dir = self.profile_dir(username)
        if profile_dir.exists():
            shutil.rmtree(profile_dir, ignore_errors=True)
            print(f"[Storage] Deleted files for @{username}")

    # ------------------------------------------------------------------
    # Storage size
    # ------------------------------------------------------------------

    def downloaded_size(self, username: str) -> int:
        """Return total bytes downloaded for a profile."""
        total = 0
        pdir = self.profile_dir(username)
        if not pdir.exists():
            return 0
        for root, _, files in os.walk(pdir):
            for fname in files:
                try:
                    total += os.path.getsize(os.path.join(root, fname))
                except OSError:
                    pass
        return total

    # ------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------

    @staticmethod
    def format_bytes(n: int) -> str:
        for unit in ("B", "KB", "MB", "GB", "TB"):
            if n < 1024:
                return f"{n:.1f} {unit}" if unit != "B" else f"{n} B"
            n /= 1024
        return f"{n:.1f} PB"

File: InstaArchive/Services/StorageManager.swift
import Foundation

/// Manages file storage for downloaded media and app data persistence
class StorageManager {
    static let shared = StorageManager()

    private let fileManager = FileManager.default
    private let settings = AppSettings.shared
    private let profilesFileName = "profiles.json"
    private let mediaIndexFileName = "media_index.json"
    private let knownMediaExtensions = ["mp4", "jpg", "jpeg", "png", "webp"]

    private var appSupportDirectory: URL {
        let appSupport = fileManager.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        return appSupport.appendingPathComponent("InstaArchive")
    }

    private init() {
        ensureDirectoryExists(appSupportDirectory)
    }

    // MARK: - Directory Management

    /// Ensure a directory exists, creating it if necessary
    @discardableResult
    func ensureDirectoryExists(_ url: URL) -> Bool {
        if !fileManager.fileExists(atPath: url.path) {
            do {
                try fileManager.createDirectory(at: url, withIntermediateDirectories: true)
                return true
            } catch {
                print("Failed to create directory at \(url.path): \(error)")
                return false
            }
        }
        return true
    }

    /// Create the directory structure for a profile
    func createProfileDirectories(for username: String) {
        let profileDir = settings.profileDirectory(for: username)
        ensureDirectoryExists(profileDir)

        for mediaType in MediaType.allCases {
            let typeDir = settings.mediaDirectory(for: username, type: mediaType)
            ensureDirectoryExists(typeDir)
        }
    }

    /// Get the save path for a media item
    func savePath(
        for media: DiscoveredMedia,
        username: String,
        index: Int = 0,
        mediaURL: String? = nil,
        fileExtension: String? = nil
    ) -> URL {
        let dir = settings.mediaDirectory(for: username, type: media.mediaType)
        let filename = baseFileName(for: media, index: index)
        let ext = (fileExtension ?? preferredFileExtension(for: mediaURL, fallbackIsVideo: media.isVideo))
            .lowercased()
        return dir.appendingPathComponent("\(filename).\(ext)")
    }

    /// Find an already-downloaded path for a media item regardless of extension.
    func existingSavePath(for media: DiscoveredMedia, username: String, index: Int = 0, mediaURL: String? = nil) -> URL? {
        let preferred = savePath(for: media, username: username, index: index, mediaURL: mediaURL)
        if fileManager.fileExists(atPath: preferred.path) {
            return preferred
        }

        let dir = settings.mediaDirectory(for: username, type: media.mediaType)
        let baseName = baseFileName(for: media, index: index)
        for ext in knownMediaExtensions {
            let candidate = dir.appendingPathComponent("\(baseName).\(ext)")
            if fileManager.fileExists(atPath: candidate.path) {
                return candidate
            }
        }
        return nil
    }

    /// Infer the expected file extension from the media URL before download.
    func preferredFileExtension(for mediaURL: String?, fallbackIsVideo: Bool) -> String {
        if let ext = fileExtensionFromURL(mediaURL) {
            return ext
        }
        return fallbackIsVideo ? "mp4" : "jpg"
    }

    /// Infer the final file extension from the downloaded bytes, then fall back to URL hints.
    func detectedFileExtension(for data: Data, mediaURL: String?, fallbackIsVideo: Bool) -> String {
        if isMP4(data) { return "mp4" }
        if isJPEG(data) { return "jpg" }
        if isPNG(data) { return "png" }
        if isWEBP(data) { return "webp" }
        return preferredFileExtension(for: mediaURL, fallbackIsVideo: fallbackIsVideo)
    }

    private func baseFileName(for media: DiscoveredMedia, index: Int) -> String {
        let timestamp = Int(media.timestamp.timeIntervalSince1970)
        let suffix = media.mediaURLs.count > 1 ? "_\(index + 1)" : ""
        return "\(media.instagramId)_\(timestamp)\(suffix)"
    }

    private func fileExtensionFromURL(_ mediaURL: String?) -> String? {
        guard let mediaURL, let url = URL(string: mediaURL) else { return nil }
        let pathExt = url.pathExtension.lowercased()
        if !pathExt.isEmpty {
            switch pathExt {
            case "jpeg": return "jpg"
            case "jpg", "png", "webp", "mp4", "mov", "m4v":
                return pathExt == "mov" || pathExt == "m4v" ? "mp4" : pathExt
            default:
                break
            }
        }

        let lowered = mediaURL.lowercased()
        if lowered.contains(".mp4") || lowered.contains("video") {
            return "mp4"
        }
        if lowered.contains(".webp") {
            return "webp"
        }
        if lowered.contains(".png") {
            return "png"
        }
        return nil
    }

    private func isJPEG(_ data: Data) -> Bool {
        data.count >= 3 && data.prefix(3) == Data([0xFF, 0xD8, 0xFF])
    }

    private func isPNG(_ data: Data) -> Bool {
        data.count >= 8 && data.prefix(8) == Data([0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A])
    }

    private func isWEBP(_ data: Data) -> Bool {
        guard data.count >= 12 else { return false }
        return String(data: data.prefix(4), encoding: .ascii) == "RIFF"
            && String(data: data.subdata(in: 8..<12), encoding: .ascii) == "WEBP"
    }

    private func isMP4(_ data: Data) -> Bool {
        guard data.count >= 12 else { return false }
        return String(data: data.subdata(in: 4..<8), encoding: .ascii) == "ftyp"
    }

    // MARK: - Profile Persistence

    /// Save profiles to disk
    func saveProfiles(_ profiles: [Profile]) {
        let url = appSupportDirectory.appendingPathComponent(profilesFileName)
        do {
            let data = try JSONEncoder().encode(profiles)
            try data.write(to: url, options: .atomicWrite)
        } catch {
            print("Failed to save profiles: \(error)")
        }
    }

    /// Load profiles from disk
    func loadProfiles() -> [Profile] {
        let url = appSupportDirectory.appendingPathComponent(profilesFileName)
        guard fileManager.fileExists(atPath: url.path) else { return [] }

        do {
            let data = try Data(contentsOf: url)
            return try JSONDecoder().decode([Profile].self, from: data)
        } catch {
            print("Failed to load profiles: \(error)")
            return []
        }
    }

    // MARK: - Media Index Persistence

    /// Save downloaded media index to disk
    func saveMediaIndex(_ items: [MediaItem]) {
        let url = appSupportDirectory.appendingPathComponent(mediaIndexFileName)
        do {
            let data = try JSONEncoder().encode(items)
            try data.write(to: url, options: .atomicWrite)
        } catch {
            print("Failed to save media index: \(error)")
        }
    }

    /// Load downloaded media index from disk
    func loadMediaIndex() -> [MediaItem] {
        let url = appSupportDirectory.appendingPathComponent(mediaIndexFileName)
        guard fileManager.fileExists(atPath: url.path) else { return [] }

        do {
            let data = try Data(contentsOf: url)
            return try JSONDecoder().decode([MediaItem].self, from: data)
        } catch {
            print("Failed to load media index: \(error)")
            return []
        }
    }

    // MARK: - File Operations

    /// Save data to a file path
    func saveFile(data: Data, to url: URL) throws {
        ensureDirectoryExists(url.deletingLastPathComponent())
        try data.write(to: url, options: .atomicWrite)
    }

    /// Check if a media item has already been downloaded
    func isMediaDownloaded(instagramId: String, in items: [MediaItem]) -> Bool {
        items.contains { $0.instagramId == instagramId }
    }

    /// Delete all downloaded files for a profile
    func deleteProfileFiles(for username: String) {
        let profileDir = settings.profileDirectory(for: username)
        do {
            if fileManager.fileExists(atPath: profileDir.path) {
                try fileManager.removeItem(at: profileDir)
                print("[InstaArchive] Deleted files for @\(username) at \(profileDir.path)")
            }
        } catch {
            print("[InstaArchive] Failed to delete files for @\(username): \(error)")
        }
    }

    /// Get total size of downloaded files for a profile
    func downloadedSize(for username: String) -> Int64 {
        let profileDir = settings.profileDirectory(for: username)
        return directorySize(at: profileDir)
    }

    /// Calculate total size of a directory recursively
    private func directorySize(at url: URL) -> Int64 {
        guard let enumerator = fileManager.enumerator(
            at: url,
            includingPropertiesForKeys: [.fileSizeKey],
            options: [.skipsHiddenFiles]
        ) else { return 0 }

        var totalSize: Int64 = 0
        for case let fileURL as URL in enumerator {
            guard let resourceValues = try? fileURL.resourceValues(forKeys: [.fileSizeKey]),
                  let fileSize = resourceValues.fileSize else { continue }
            totalSize += Int64(fileSize)
        }
        return totalSize
    }

    /// Format bytes into human-readable string
    static func formatBytes(_ bytes: Int64) -> String {
        let formatter = ByteCountFormatter()
        formatter.countStyle = .file
        return formatter.string(fromByteCount: bytes)
    }
}
